### PR TITLE
Set up resources in env/project stack for DNS Delegation

### PR DESCRIPTION
**Pull request by kohidave**
This pipes through the DNS name of the project to the env
templates, and wires in the custom resources.

TODO: Add the integration tests for this - but the
existing, non DNS integ tests are running fine.

I manually ran the env setup for a DNS project. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: internal/pkg/deploy/cloudformation/cloudformation_integration_test.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// +build integration
+// build integration
```
Comment (efekarakus) [inline internal/pkg/deploy/cloudformation/cloudformation_integration_test.go]: huh what's the difference? why did we remove it

File: internal/pkg/deploy/cloudformation/stack/env.go [hunk 2]
```diff
@@ -49,11 +55,41 @@ func NewEnvStackConfig(input *deploy.CreateEnvironmentInput, box packd.Box) *Env
...
+		return "", err
+	}
+
+	templateData := struct {
+		DNSDelegationLambda string
+		ACMValidationLambda string
+	}{
+		dnsDelegator,
+		acmValidator,
+	}
```
Comment (efekarakus) [inline internal/pkg/deploy/cloudformation/stack/env.go]: Beautiful 😅

File: internal/pkg/deploy/cloudformation/stack/env.go [hunk 3]
```diff
@@ -31,6 +35,8 @@ const (
 	envParamProjectNameKey           = "ProjectName"
 	envParamEnvNameKey               = "EnvironmentName"
 	envParamToolsAccountPrincipalKey = "ToolsAccountPrincipalARN"
+	envParamProjectDNSKey            = "ProjectDNSName"
+	envParamProjectDNSDelegationRole = "ProjectDNSDelegationRole"
```
Comment (efekarakus) [inline internal/pkg/deploy/cloudformation/stack/env.go]: nit: Suffix const name with "Key"

File: templates/environment/cf.yml [hunk 4]
```diff
@@ -437,6 +470,106 @@ Resources:
...
+    Type: AWS::Lambda::Function
+    Condition: DelegateDNS
+    Properties:
+      Code:
+        ZipFile: |
+          {{.ACMValidationLambda}}
+      Handler: "index.certificateRequestHandler"
+      Timeout: 600
+      Role: !GetAtt 'CustomResourceRole.Arn'
+      Runtime: nodejs10.x
```
Comment (efekarakus) [inline templates/environment/cf.yml]: nit: They announced 12.x!

File: templates/environment/cf.yml [hunk 5]
```diff
@@ -437,6 +470,106 @@ Resources:
             ]
             Resource: "*"
 
+  # DNS Delegation Resources
+  CertificateValidationFunction:
```
Comment (efekarakus) [inline templates/environment/cf.yml]: Can you check from the logs how much memory the fn is using now? To make sure we're not close to 128

File: templates/environment/cf.yml [hunk 6]
```diff
@@ -437,6 +470,106 @@ Resources:
...
+                  - "sts:AssumeRole"
+                  - "logs:*"
+                  - "route53:ChangeResourceRecordSets"
+                  - "route53:ListResourceRecordSets"
+                  - "route53:Get*"
+                  - "route53:Describe*"
+                Resource:
+                  - "*"
+  # Delegates control of our hosted zone in the project's domain
+  # and
```
Comment (efekarakus) [inline templates/environment/cf.yml]: And??

File: templates/environment/cf.yml [hunk 7]
```diff
@@ -437,6 +470,106 @@ Resources:
...
+          -
+            Effect: Allow
+            Principal:
+              Service:
+                - lambda.amazonaws.com
+            Action:
+              - sts:AssumeRole
+      Path: /
+      Policies:
+        - PolicyName: "lambda-logs"
```
Comment (efekarakus) [inline templates/environment/cf.yml]: nit: should we rename this policy?

File: internal/pkg/deploy/cloudformation/cloudformation_integration_test.go (hunk 1 above)
Comment (kohidave) [reply internal/pkg/deploy/cloudformation/cloudformation_integration_test.go]: Whoops. That's why the tests were failing!

File: templates/environment/cf.yml [hunk 8]
```diff
@@ -437,6 +470,109 @@ Resources:
...
+      NameServers: !GetAtt EnvironmentHostedZone.NameServers
+      RootDNSRole: !Ref ProjectDNSDelegationRole
+
+  HTTPSCert:
+    Condition: DelegateDNS
+    Type: Custom::CertificateValidationFunction
+    DependsOn:
+    - CertificateValidationFunction
+    - EnvironmentHostedZone
+    - DelegateDNSAction
```
Comment (efekarakus) [inline templates/environment/cf.yml]: Does this force the `CertificateValidationFunction` to run after creating the `DNSDelegationFunction`?

Comment (kohidave) [reply templates/environment/cf.yml]: It forces this to wait until DNS delegation is doen.

**Pull request merged**